### DISCUSSION TITLE
Normalize weights

### DISCUSSION THREAD

**Discussion by JTriggerFish** [ljleb/sd-mecha#46] [Q&A]
Thanks for the awesome library !
Is there a way to normalize the layers of a checkpoint after merging, say to keep their standard deviation ( not sure accross what dimensions makes the most sense ) close to 1, or to moment match the standard deviations of the layers of another checkpoint ?
Some SDXL merges end up very oversaturated and I'm pretty sure it's due to lack of normalization.

**Comment by ljleb** [comment]
Hi, oversaturation usually is caused by add_difference or other similar methods that do not strictly interpolate. This is usually because the parameters of the model land outside of a range that is sensible for the architecture. As you note, it is probably related to variance collapse (in this case, variance explosion I guess)

To remedy this, there is a technique know as weights clipping regularization. Basically, you take two models $\theta_0$ and $\theta_1$, and for each pair of parameters $\theta_0^i$ and $\theta_1^i$, force the $i$-th parameter of the merge to lie within $[min(\theta_0^i, \theta_1^i), max(\theta_0^i, \theta_1^i)]$. `sd_mecha.clamp(a, *bounds)` implements this: `a` is the model to regularize, and `*bounds` should be 2 models of your choice (usually they are input models that were used to obtain clamp's `a`, i.e. the `a` and `b` passed to add difference).

You can use more than 2 models as boundaries if you want, in which case it will relax the range to $[min(\theta_0^i, \theta_1^i, \theta_2^i, ...), max(\theta_0^i, \theta_1^i, \theta_2^i, ...)]$ (this is usually undesirable, but it can be good to know when clipping to 2 models is too strict)

Note that while this works quite well in some scenarios, if your input models are simply too far apart (i.e. PonyDiffusion 6 and Animagine 4) weights clipping will end up preferring one model over the other. It may not fully fix the merge either, although usually it does a fine job.

Hope this helps!

**Comment by ljleb** [comment]
As a side note, you may want to take a look at this paper, which should cover a similar case to yours. Although they focus on variance collapse instead of variance explosion, I think a similar idea can still be applied to restore variance to an acceptable scale.

https://arxiv.org/abs/2211.08403

They appear to show that you cannot do this very well without a dataset, however their method also appears to have a surprisingly good performance.

Somewhat related to this, an alternative idea I explored is to rescale each linear layer of a merged model so that the pseudo-determinant is equal to the geometric average of the pseudo-determinants of the corresponding linear layers in two other input models. The pseudo-determinant is a measure of how a linear transform scales its input.

Let $\theta_0^w$ and $\theta_1^w$ be the corresponding linear layers of two input models, and $\theta_{\text{out}}^w$ be the corresponding linear layer for the unnormalized merge or output model. Note that $\theta_k^w \in \mathbb{R}^{p \times q}$ (and $k = 0, 1, \text{out}$) -- they are the linear transforms in the architecture. For convolution layers (that are obviously not square), it is possible to reshape them from $\mathbb{R}^{p \times q \times h \times w}$ to $\mathbb{R}^{p \times (qhw)}$, or to simply weighted-average them (although I think there might be a better way to handle conv layers). All other layers are simply weighted-averaged. The pseudo-determinant of $\theta_k^w$ can be computed as

$$
\text{pdet}(\theta_k^w) = exp(\frac{1}{n} \sum_{i=1}^n log(\sigma_k^i))
$$

where $\theta_k^w = U_k \text{diag}(\sigma_k) V_k^T$ is the singular value decomposition of $\theta_k^w$. $\sigma_k^i$ is a singular value of the linear transform for $i = 1, 2, ..., n$ and $n = min(p, q)$.

Now, let $\theta_{\text{normalized}}^w = \frac{\theta_{\text{out}}^w \times \sqrt{\text{pdet}(\theta_0^w) \times \text{pdet}(\theta_1^w)}}{\text{pdet}(\theta_{\text{out}}^w)}$ be the merged layer with its pseudo-determinant replaced by the geometric average of that of $\theta_0^w$ and $\theta_1^w$. If all models are not too far apart, it actually works surprisingly well.

**Comment by JTriggerFish** [comment]
Thanks very much for this ! I will take a look at the paper.

Yes it’s indeed add difference giving me issues.
I’ve found add difference to be a good way to transfer knowledge from two very different models by doing A + alpha * (B-SDXL base), which in my experience works better than doing a weighted sum of A and B ( which works fine for more similar models ) but this issue can happen.
I’ve used the clamping on occasions but found it can give ‘weirder’ results - probably as you say due to some layers ending up preferring one or the other models.
Handling different layers differently and the pseudo determinant are very interesting ideas, will look into this.

**Reply by ljleb** [reply]
If you're going to attempt pseudo-determinant rescale, note that although it can work surprisingly well, sometimes it does not really fix the underlying issues. It really depends on the input models. I think there is no paper about this method probably because it isn't a super justifiable approach, it's just a heuristic that sometimes works.

**Reply by ljleb** [reply]
Since you are using add_difference, have you tried train_difference or add_opposite by the way? These are similar to add_difference but they try to mitigate the scale issue by not adding parameters that are already similar to each other. You could have more luck with these.

**Reply by ljleb** [reply]
Here's an explanation for add opposite in case you could be interested in this: https://github.com/hako-mikan/sd-webui-supermerger/discussions/264#discussioncomment-8639803

**Reply by JTriggerFish** [reply]
Super interesting, I wasn’t actually aware of these I will take a look, thank you !

**Reply by JTriggerFish** [reply]
@ljleb in order to build intuition into what merging strategy works best I’d like to compute statistics such as some mean, median and quantiles of std deviation of each layer of a given checkpoint, possibly organised as a multi indexed pandas df indexed by different parts of the network etc.
This way I can have some objective metrics since exhaustive testing through image generation is very hard.
If you had any tips or code snippets to get me started would be much appreciated !
Happy to contribute the code back as a merge request if it’s useful.

**Reply by ljleb** [reply]
I like this idea! Usually I use ComfyUI to gather information from inference. I patch the code by hand and then save safetensors along the way. I don't really recommend this to be honest since you end up with all your experiments in the same repo, it gets clunky and error-prone.

I think the diffusers library could be a better way to do this. Consider loading any checkpoint for inference, then using pytorch forward hooks you can gather activation stats, and then you can call the pipeline to generate images. I suppose the statistics of attention maps from the attention layers (the product post softmax, $\text{Softmax}(Q K^T)$ ) could provide the most informative stats, however you are free to look at other activations, this is just a suggestion.

**Reply by ljleb** [reply]
I went ahead and created this in a back-and-forth with o3-mini:

code: https://gist.github.com/ljleb/776a4ef78a8bc1a2f6a411710eb6c920
logs: https://gist.github.com/ljleb/f5a1f9cf2aa67aab9fa78a630deabee2

I realize now that since this is post-softmax, the mean, max and min are probably not very informative. I guess other activations could carry more information about variance collapse, like the output of the entire attention blocks. The variance does appear to vary so the main metric we are interested in is there at least. Visualization is not very good yet, it just prints a bunch of numbers.

**Reply by JTriggerFish** [reply]
That's pretty cool ! What I had in mind was actually stats on the weights themselves rather than stats on the outputs of each layer during the forward passes - though that could be very useful too.

**Reply by JTriggerFish** [reply]
I'm not really sure where to start in terms of loading a checkpoint ( either from disk or for a merge ) and iterating through the layers to compute stats on them as the sd_mecha code is quite abstracted ( and very efficient it seems ) so any pointers would be great, but otherwise I'll try to step through the code with the debugger

**Reply by JTriggerFish** [reply]
Well that was actually easier than I thought and sd_mecha would probably have been overkill for this, I got o1 to write me this which is a bit rough around the edges but does the job:

https://gist.github.com/JTriggerFish/d95d30d260b14882b57a16a655d95b4a

**Reply by ljleb** [reply]
Ah I see, I misinterpreted your suggestion. Well, all is good if you got around it.

If you are interested in gathering stats on the weights themselves with sd-mecha, it's actually fairly easy. If you're on rc1, you can write a custom merge method like this:

```python
import sd_mecha
import torch
from sd_mecha import merge_method, Parameter, Return
from torch import Tensor

@merge_method
def stats(a: Parameter(Tensor)) -> Return(Tensor):
    return torch.cat([a.mean(), a.min(), a.max(), a.std()])

# and then we can use it:
merger = sd_mecha.RecipeMerger()
output = {}
merger.merge_and_save(stats(sd_mecha.model("path/to/model")), output)

# output now contains the stats for each key
```

(edit: as a side note, the RecipeMerger class was made with bulk merging in mind, where reusing defaults makes sense. In rc2, merging a simple model can also be done through `sd_mecha.merge_and_save` which would save 2 lines of code here)

**Reply by JTriggerFish** [reply]
oh wow that is much nicer, great I will give that a go ! Thanks again.

**Reply by ljleb** [reply]
Awesome! I suggest giving this a read if you ever intend to work with the lib, it should explain quite a lot:

https://github.com/ljleb/sd-mecha/tree/real-docs/guide

Feel free to skip over any section that doesn't do much for you if you do give it a read. It is a work in progress documentation that covers current features and one or two features I intend to add in rc2. (namely, `sd_mecha.merge_and_save` and `sd_mecha.pick_component`)
